include/experimental/kernel/col_major_shuf/col_major_shuf_xe.hpp

```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
+ * Copyright (c) 2023-2024 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+/// @file
+/// C++ API
+
+#pragma once
+
+#include <experimental/kernel/col_major_shuf/api.hpp>
+#include <experimental/kernel/col_major_shuf/common.hpp>
+#include <experimental/kernel/col_major_shuf/config.hpp>
+
+namespace gpu::xetla::kernel {
+template <
+    typename dtype_in_,
+    typename dtype_out_,
+    typename dtype_gidx_,
+    mem_layout mem_layout_in_,
+    typename col_major_shuf_attr_,
+    gpu_arch arch_>
+struct col_major_shuf_t<
+    dtype_in_,
+    dtype_out_,
+    dtype_gidx_,
+    mem_layout_in_,
+    col_major_shuf_attr_,
+    arch_,
+    std::enable_if_t<arch_ == gpu_arch::XeHpg || arch_ == gpu_arch::XeLpg>> {
+  using dtype_in = dtype_in_;
+  using dtype_out = dtype_out_;
+  using dtype_gidx = dtype_gidx_;
+  using col_major_shuf_attr = col_major_shuf_attr_;
+
+  static constexpr mem_layout mem_layout_in = mem_layout_in_;
+
+  static_assert(
+      std::is_same<dtype_in, dtype_out>::value,
+      "only support in/out data type must be same now.");
+  static_assert(
+      mem_layout_in == mem_layout::row_major,
+      "only support row_major input now.");
+  static_assert(
+      std::is_same<dtype_gidx, uint32_t>::value,
```

Comment:
Is int8 not allowed?

Reply:
yes, if the k dim larger than 2^8 then int8 dtype can't work,
in fact, it should always be uint32_t.